File: apps/android/app/src/main/java/com/example/creditcard/utils/WalletCardRules.kt
package com.example.creditcard.utils

import com.example.creditcard.ui.main.getCardBrand
import com.example.creditcard.data.SharedCard
import java.math.BigDecimal
import java.math.RoundingMode
import java.time.LocalDate
import java.time.temporal.ChronoUnit
import java.util.Locale

/** Contract v1; no UI state and no changes to the SyncV4 wire model. */
object WalletCardRules {
    data class CreditPool(val country: String, val bank: String, val currency: String)
    fun currency(card: SharedCard): String = card.type.trim().uppercase(Locale.ROOT)
    fun pool(card: SharedCard) = CreditPool(card.country.trim(), normalizeBankNameForMatch(card.bank), currency(card))
    private fun cents(value: Double): BigDecimal = if (value.isFinite() && value >= 0)
        BigDecimal.valueOf(value).movePointRight(2).setScale(0, RoundingMode.HALF_UP) else BigDecimal.ZERO

    fun creditLimits(cards: List<SharedCard>): Map<String, Double> {
        val shared = mutableMapOf<CreditPool, BigDecimal>()
        val totals = mutableMapOf<String, BigDecimal>()
        cards.filter { it.cardCategory != "debit" }.forEach { card ->
            val key = pool(card)
            val amount = cents(card.limit)
            if (card.isSharedLimit && key.bank.isNotEmpty()) {
                shared[key] = (shared[key] ?: BigDecimal.ZERO).max(amount)
            } else totals[key.currency] = (totals[key.currency] ?: BigDecimal.ZERO) + amount
        }
        shared.forEach { (key, amount) -> totals[key.currency] = (totals[key.currency] ?: BigDecimal.ZERO) + amount }
        return totals.mapValues { it.value.movePointLeft(2).toDouble() }.toSortedMap()
    }

    fun interestFreeDays(card: SharedCard, today: LocalDate = LocalDate.now()): Int {
        if (card.cardCategory == "debit") return -1
        val bill = card.accountBillDate.trim().toIntOrNull() ?: return -1
        val due = card.dueDate.trim().toIntOrNull() ?: return -1
        if (bill !in 1..31 || due !in 1..31) return -1
        val actualBill = minOf(bill, today.lengthOfMonth())
        val next = if (card.billingDaySpendingToNextBill) today.dayOfMonth >= actualBill else today.dayOfMonth > actualBill
        val month = today.withDayOfMonth(1).plusMonths(if (next) 1 else 0)
        val dueMonth = month.plusMonths(if (due <= bill) 1 else 0)
        val target = dueMonth.withDayOfMonth(minOf(due, dueMonth.lengthOfMonth()))
        return ChronoUnit.DAYS.between(today, target).toInt().coerceAtLeast(0)
    }

    fun searchText(card: SharedCard): String = listOf(card.bank, card.alias, card.cardNumber,
        card.level, card.type, card.country, card.equity, card.remark, card.limit.toString(),
        if (card.cardCategory == "debit") "储蓄卡 儲蓄卡 debit" else "信用卡 credit",
        getCardBrand(card.cardNumber, card.level)).joinToString("\n").lowercase(Locale.ROOT)

    fun matches(card: SharedCard, query: String, index: String = searchText(card)): Boolean {
        val text = query.trim().lowercase(Locale.ROOT)
        if (text.isEmpty()) return true
        val compact = text.replace(Regex("[\\s-]"), "")
        if (compact.isEmpty()) return false
        val number = card.cardNumber.replace(Regex("[\\s-]"), "")
        return index.contains(text) || (compact.all { it in '0'..'9' } && number.contains(compact))
    }

    fun sorted(cards: List<SharedCard>, key: String, today: LocalDate = LocalDate.now()): List<SharedCard> {
        val days = if (key.startsWith("interest-")) cards.associate { it.id to interestFreeDays(it, today) } else emptyMap()
        return cards.sortedWith { a, b ->
            val order = when (key) {
                "limit-desc", "limit-asc" -> {
                    val left = if (a.cardCategory == "debit") 0.0 else a.limit.coerceAtLeast(0.0)
                    val right = if (b.cardCategory == "debit") 0.0 else b.limit.coerceAtLeast(0.0)
                    if (key == "limit-asc") left.compareTo(right) else right.compareTo(left)
                }
                "interest-desc", "interest-asc" -> {
                    val left = days.getValue(a.id); val right = days.getValue(b.id)
                    when {
                        left < 0 && right >= 0 -> 1
                        left >= 0 && right < 0 -> -1
                        key == "interest-asc" -> left.compareTo(right)
                        else -> right.compareTo(left)
                    }
                }
                "modifyTime" -> b.lastModifyTime.compareTo(a.lastModifyTime)
                "bank-asc" -> compareValuesBy(a, b, { it.bank }, { it.alias })
                else -> compareValuesBy(a, b, { it.country }, { it.bank }, { it.alias })
            }
            if (order != 0) order else a.id.compareTo(b.id)
        }
    }
}

File: apps/macos/Domain/WalletCardRules.swift
import Foundation

/// Contract v1: currency-separated credit pools and Gregorian civil-day calculations.
public enum WalletCardRules {
    public struct CreditPool: Hashable, Sendable {
        public let country: String
        public let bank: String
        public let currency: String
    }

    public static func currency(_ card: SharedCard) -> String {
        (card.type ?? "").trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
    }

    public static func pool(_ card: SharedCard) -> CreditPool {
        CreditPool(country: card.country.trimmingCharacters(in: .whitespacesAndNewlines),
                   bank: BankNameNormalizer.normalizedKey(card.bank), currency: currency(card))
    }

    private static func cents(_ value: Double?) -> Decimal {
        guard let value, value.isFinite, value >= 0 else { return 0 }
        var amount = Decimal(string: String(value), locale: Locale(identifier: "en_US_POSIX")) ?? 0
        amount *= 100
        var result = Decimal()
        NSDecimalRound(&result, &amount, 0, .plain)
        return result
    }

    public static func creditLimits(_ cards: [SharedCard]) -> [String: Double] {
        var shared: [CreditPool: Decimal] = [:]
        var totals: [String: Decimal] = [:]
        for card in cards where card.cardCategory != "debit" {
            let key = pool(card), amount = cents(card.limit)
            if card.isSharedLimit && !key.bank.isEmpty {
                shared[key] = max(shared[key] ?? 0, amount)
            } else {
                totals[key.currency, default: 0] += amount
            }
        }
        for (key, amount) in shared { totals[key.currency, default: 0] += amount }
        return totals.mapValues { NSDecimalNumber(decimal: $0 / 100).doubleValue }
    }

    /// Remaining interest-free days for a purchase today; invalid/debit cards return -1.
    public static func interestFreeDays(_ card: SharedCard, today: Date = Date(), timeZone: TimeZone = .current) -> Int {
        guard card.cardCategory != "debit",
              let bill = Int((card.accountBillDate ?? "").trimmingCharacters(in: .whitespacesAndNewlines)),
              let due = Int((card.dueDate ?? "").trimmingCharacters(in: .whitespacesAndNewlines)),
              (1...31).contains(bill), (1...31).contains(due) else { return -1 }
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = timeZone
        let day = cal.startOfDay(for: today)
        guard let month = cal.date(from: cal.dateComponents([.year, .month], from: day)),
              let range = cal.range(of: .day, in: .month, for: month) else { return -1 }
        let actualBill = min(bill, range.count), spend = cal.component(.day, from: day)
        let next = card.billingDaySpendingToNextBill ? spend >= actualBill : spend > actualBill
        guard let billMonth = cal.date(byAdding: .month, value: next ? 1 : 0, to: month),
              let dueMonth = cal.date(byAdding: .month, value: due <= bill ? 1 : 0, to: billMonth),
              let dueRange = cal.range(of: .day, in: .month, for: dueMonth),
              let target = cal.date(byAdding: .day, value: min(due, dueRange.count) - 1, to: dueMonth) else { return -1 }
        return max(0, cal.dateComponents([.day], from: day, to: target).day ?? 0)
    }
}
